data/src/client.rs

```diff
@@ -679,35 +698,68 @@ impl Client {
                             log::debug!("[{}] {target} - WHO receiving...", self.server);
                         }
 
-                        // H = Here, G = gone (away)
-                        let flags = args.get(6)?.chars().collect::<Vec<char>>();
-                        let away = *(flags.first()?) == 'G';
-
-                        let lookup = User::from(Nick::from(args[5].clone()));
-
-                        if let Some(mut user) = channel.users.take(&lookup) {
-                            user.update_away(away);
-                            channel.users.insert(user);
-                        }
+                        channel.update_user_away(
+                            User::from(Nick::from(args.get(5)?.clone())),
+                            args.get(6)?.chars().collect::<Vec<char>>(),
+                        );
 
                         // We requested, don't save to history
-                        if matches!(channel.last_who, Some(WhoStatus::Receiving)) {
+                        if matches!(channel.last_who, Some(WhoStatus::Receiving(None))) {
                             return None;
                         }
                     }
                 }
             }
+            Command::Numeric(RPL_WHOSPCRPL, args) => {
+                let target = args.get(2)?;
+
+                if proto::is_channel(target) {
+                    if let Some(channel) = self.chanmap.get_mut(target) {
+                        if let Ok(token) = args.get(1)?.parse::<isupport::WhoToken>() {
+                            if let Some(WhoStatus::Requested(_, Some(request_token))) =
+                                channel.last_who
+                            {
+                                if request_token == token {
+                                    channel.last_who =
+                                        Some(WhoStatus::Receiving(Some(request_token)));
+                                    log::debug!("[{}] {target} - WHO receiving...", self.server);
+                                }
+                            }
+
+                            // We requested, don't save to history
+                            if let Some(WhoStatus::Receiving(Some(request_token))) =
+                                channel.last_who
+                            {
+                                if request_token == token {
+                                    let flags = args.get(4)?.chars().collect::<Vec<char>>();
+
+                                    if flags
+                                        .iter()
+                                        .all(|c| matches!(c, 'H' | 'G' | '&' | '@' | '%' | '+'))
+                                    {
+                                        if let Ok(user) = User::try_from(args.get(3)?.clone()) {
+                                            channel.update_user_away(user, flags);
+                                        }
+                                    }
```

Comment:
Whys this logic different than in `RPL_WHOREPLY`? Also, we don't need to have the logic here and do the token == token logic again, just put this up after line `730` like

Reply:
I added a bit more parameter checking to this call because the user could manually invoke a `WHOX` command with different tags specified than those used in `WHOX` polling.  I left it out of the `WHO` polling as a minor optimization, but we could definitely fold it into `update_user_away` for increased code simplicity.  Should I do that?

The `request_token == token` check above is for `WhoStatus::Requested`, so I believe it will only be matched for the very first `RPL_WHOSPCRPL` received.  I attempted to model that after how `RPL_WHOREPLY` handles its messages.

Reply:
After changing the `flags` parsing (couldn't reasonably make the check more stringent for `RPL_WHOSPCRPL` after looking further into what `flags` is allowed to contain), there really wasn't any reason to separate the two.  Went ahead and consolidated further.  Thanks for the suggestion.